testing/src/main/java/io/grpc/internal/testing/AbstractTransportTest.java

```diff
@@ -0,0 +1,907 @@
+  //     (and maybe exceptions handled)
+
+  @Test
+  public void serverNotListening() {
+    server = null;
+    InOrder inOrder = inOrder(mockClientTransportListener);
+    client.start(mockClientTransportListener);
+    verify(mockClientTransportListener, timeout(1000)).transportTerminated();
+    inOrder.verify(mockClientTransportListener).transportShutdown(statusCaptor.capture());
+    assertCodeEquals(Status.UNAVAILABLE, statusCaptor.getValue());
+    inOrder.verify(mockClientTransportListener).transportTerminated();
+    verify(mockClientTransportListener, never()).transportReady();
+  }
+
+  @Test
+  public void clientStartStop() throws Exception {
+    server.start(serverListener);
+    InOrder inOrder = inOrder(mockClientTransportListener);
+    client.start(mockClientTransportListener);
+    client.shutdown();
+    verify(mockClientTransportListener, timeout(1000)).transportTerminated();
+    inOrder.verify(mockClientTransportListener).transportShutdown(any(Status.class));
+    inOrder.verify(mockClientTransportListener).transportTerminated();
+    server.shutdown();
+    assertTrue(serverListener.waitForShutdown(1000, TimeUnit.MILLISECONDS));
+    server = null;
+  }
+
+  @Test
+  public void clientStartAndStopOnceConnected() throws Exception {
+    server.start(serverListener);
+    InOrder inOrder = inOrder(mockClientTransportListener);
+    client.start(mockClientTransportListener);
+    verify(mockClientTransportListener, timeout(1000)).transportReady();
+    MockServerTransportListener serverTransportListener
+        = serverListener.takeListenerOrFail(1000, TimeUnit.MILLISECONDS);
+    client.shutdown();
+    verify(mockClientTransportListener, timeout(1000)).transportTerminated();
+    inOrder.verify(mockClientTransportListener).transportShutdown(any(Status.class));
+    inOrder.verify(mockClientTransportListener).transportTerminated();
+    assertTrue(serverTransportListener.waitForTermination(1000, TimeUnit.MILLISECONDS));
+    server.shutdown();
+    assertTrue(serverListener.waitForShutdown(1000, TimeUnit.MILLISECONDS));
+    server = null;
+  }
+
+  @Test
+  public void serverAlreadyListening() throws Exception {
+    client = null;
+    server.start(serverListener);
+    Server server2 = newServer();
+    thrown.expect(IOException.class);
+    server2.start(new MockServerListener());
+  }
+
+  @Test
+  public void openStreamPreventsTermination() throws Exception {
+    server.start(serverListener);
+    client.start(mockClientTransportListener);
+    MockServerTransportListener serverTransportListener
+        = serverListener.takeListenerOrFail(1000, TimeUnit.MILLISECONDS);
+    serverTransport = serverTransportListener.transport;
+
+    ClientStream clientStream = client.newStream(methodDescriptor, new Metadata());
+    clientStream.start(mockClientStreamListener);
+    StreamCreation serverStreamCreation
+        = serverTransportListener.takeStreamOrFail(1000, TimeUnit.MILLISECONDS);
```

Comment:
Replaced with `TIMEOUT_MS`. I actually have little problem using constants in a test, but I had actually intended to replace those before sending it out.